Add "navigator" option to "showDialog" and "showGeneralDialog"

brianegan opened the pull request:
## Description

The problem: We need to use `showDialog` to push a dialog to sub Navigator, not the root Navigator. At the moment, `showDialog` uses `showGenericDialog` which always uses the root navigator.

The solution: This Pull Request adds the ability to specify the `NavigatorState` which should be used to push dialog routes.

Alternative solution: Accept an optional `bool useRootNavigator = true` parameter instead of a `NavigatorState`.  

## Related Issues

  - Allows https://github.com/flutter/assets-for-api-docs/pull/106 to generate diagrams properly
  - Fixes https://github.com/flutter/flutter/issues/23565

## Tests

I added the following tests:

  - Before writing tests I wanted feedback on the general idea. Tests will be added after completing any necessary api discussions.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [X] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

brianegan committed:
Add `navigator` option to `showDialog` and `showGeneralDialog`

brianegan committed:
Add `navigator` to `showDatePicker`

goderbauer commented on packages/flutter/lib/src/material/dialog.dart:
```diff
@@ -657,11 +657,15 @@ Widget _buildMaterialDialogTransitions(BuildContext context, Animation<double> a
 /// widget can be safely removed from the tree before the dialog is closed.
 ///
-/// The `child` argument is deprecated, and should be replaced with `builder`.
+/// The `navigator` argument defines the [NavigatorState] to use to push the
```
This should call out that passing in a Navigator is entirely optional.

goderbauer commented on packages/flutter/lib/src/material/date_picker.dart:
```diff
@@ -1176,5 +1177,6 @@ Future<DateTime> showDatePicker({
     builder: (BuildContext context) {
       return builder == null ? child : builder(context, child);
     },
+    navigator: navigator,
```
nit: I would move this between context and builder.

goderbauer commented on packages/flutter/lib/src/widgets/routes.dart:
```diff
@@ -1533,9 +1533,13 @@ class _DialogRoute<T> extends PopupRoute<T> {
+/// dialog. It is only used when the method is called. Its corresponding widget
+/// can be safely removed from the tree before the dialog is closed.
+///
+/// The `navigator` argument defines the [NavigatorState] to use to push the
```
say that navigator is optional.

goderbauer commented on packages/flutter/lib/src/widgets/routes.dart:
```diff
@@ -1576,10 +1580,11 @@ Future<T> showGeneralDialog<T>({
   Color barrierColor,
   Duration transitionDuration,
   RouteTransitionsBuilder transitionBuilder,
+  NavigatorState navigator,
```
Looks like context doesn't have to be required anymore on this method. But we should have an assert that either context or navigator is not null.

goderbauer reviewed (commented):
Approach seems OK. This will need a test, though.

jimmy-robert reviewed (changes requested):
What about other `showXXX()` functions that also use `Navigator` or `showDialog`?
- `showAboutDialog`
- `showCupertinoDialog`
- `showCupertinoModalPopup`
- `showLicensePage`
- `showMenu`
- `showModalBottomSheet`
- `showSearch`
- `showTimePicker`

Piinks commented:
@brianegan Are you planning on revisiting this?

brianegan commented:
@Piinks Yep! Sorry, had another bit of work to wrap up the past week or so.

After seeing @JimmyRobz  comment, I realized this is actually a pretty decent sized change to quite a few methods. Since the assets-for-api-docs project is quite unique in how it works and is kind of "non-standard", I wanted to take a bit more time to see if I could get around the issue without changing so many methods in Flutter itself.

After spending a few more hours on that, I think I feel pretty confident in saying the only way to fix the issue there is to make this change!

brianegan commented:
@JimmyRobz @goderbauer Thanks for your feedback! I have a question I'd love to ask: There were two ideas for this change -- allow this function to accept a `bool useRootNavigator` argument, or allow these functions to accept a specific `NavigatorState navigator`. @goderbauer suggested we go with the `NavigatorState` option, but while going through each `showX` method, I noticed the `showModalBottomSheet` already has a `useRootNavigator` option.

### useRootNavigator

Pros:

  - API consistency with what's already in place
  - No need to remove `@required` annotation from `context` on several methods. Overall, I worry about removing that annotation and replacing it with an `assert`, since the annotation provides IDE completion / warnings whereas asserts only fail at runtime.

Cons:

  - Can't specify the *exact* navigator you want to use. Need to rely on the nearest `Navigator`.

### NavigatorState

Pros:

  - Specify the exact `NavigatorState` you want to use

Cons:

  - Requires removing `@required` annotation
  - Maybe confusing to pass one or the other option into the function? Some would require both.

goderbauer commented:
Since we already have useRootNavigator on one of them and we don't really have a user case for anything else I'd be fine with just adding useRootNavigator.

brianegan committed:
Move from `navigator` argument to `useRootNavigator`

brianegan commented:
Thanks @goderbauer -- I've now changed everything to `useRootNavigator`. I've also added 1 test to the `showDialog` test suite. Could you please verify that's the correct / best way to test this change? If it looks ok or you have improvements, I can apply a similar test to all functions.

goderbauer commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -864,6 +864,9 @@ class _CupertinoModalPopupRoute<T> extends PopupRoute<T> {
 /// safely removed from the tree before the popup is closed.
 ///
+/// The `useRootNavigator` argument is used to determine whether to push the
+/// popup to the [Navigator] furthest from or nearest to the given `context`.
```
This should mention what the default is (here and elsewhere).

goderbauer replied:
You may want to use a `{@macro ...}` for this.

goderbauer commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -933,13 +937,16 @@ Widget _buildCupertinoDialogTransitions(BuildContext context, Animation<double>
-/// If the application has multiple [Navigator] objects, it may be necessary to
-/// call `Navigator.of(context, rootNavigator: true).pop(result)` to close the
-/// dialog rather than just `Navigator.pop(context, result)`.
+/// By default, the dialog route created by this method is pushed to the root
```
This paragraph should probably come right after the paragraph that explains "useRootNavigator"

goderbauer commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -882,8 +885,9 @@ class _CupertinoModalPopupRoute<T> extends PopupRoute<T> {
+  bool useRootNavigator = true,
 }) {
-  return Navigator.of(context, rootNavigator: true).push(
+  return Navigator.of(context, rootNavigator: useRootNavigator).push(
```
Add an assert here and elsewhere that useRootNavigator is not null.

brianegan replied:
Done.

goderbauer reviewed (commented):
Looks good. Just some nits.

goderbauer commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -223,19 +223,21 @@ class AboutListTile extends StatelessWidget {
   Widget applicationIcon,
   String applicationLegalese,
   List<Widget> children,
+  bool useRootNavigator,
```
This is probably the reason for the test failures.

brianegan replied:
Fixed up!

brianegan committed:
Add tests for every function, update based on feedback

  - Tests for each function
  - Mention `useRootNavigator` default in every function's docs
  - Assert root navigator is not null in every function
  - Ensure correct default values for every function

brianegan committed:
Fix documentation

brianegan commented:
Thanks again, @goderbauer. I've added tests to each function and addressed all feedback.

  - Mention `useRootNavigator` default in every function's docs
  - Assert root navigator is not null in every function
  - Ensure correct default values for every function

goderbauer reviewed (approved):
LGTM

brianegan commented:
Thanks, @goderbauer! Should I hold off on merging this change since flutter-build is currently broken on this branch? Is there anything I can do to fix it or is the problem unrelated to my changes?

goderbauer commented:
Yeah, you'll have to wait until all checks are green. The current redness is unrelated to your change.
